From: Fabian Pflug <f.pflug@pengutronix.de>
Date: Mon, 20 Apr 2026 09:58:10 +0200
Subject: [PATCH] plat-imx: add state implementation for rpmb-pta based on
 eFuse
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add a way to lock the access to the RPMB authentication key by burning and
checking a general purpose eFuse on i.MX SoCs.

Signed-off-by: Fabian Pflug <f.pflug@pengutronix.de>
Signed-off-by: Ulrich Ölmann <u.oelmann@pengutronix.de>
---
 core/arch/arm/plat-imx/conf.mk | 19 +++++++++++++++++++
 core/arch/arm/plat-imx/main.c  | 37 +++++++++++++++++++++++++++++++++++++
 2 files changed, 56 insertions(+)

diff --git a/core/arch/arm/plat-imx/conf.mk b/core/arch/arm/plat-imx/conf.mk
index 574d9b61979c..b29619237a8a 100644
--- a/core/arch/arm/plat-imx/conf.mk
+++ b/core/arch/arm/plat-imx/conf.mk
@@ -303,6 +303,25 @@ else
 $(error Unsupported PLATFORM_FLAVOR "$(PLATFORM_FLAVOR)")
 endif
 
+ifeq ($(filter y, $(CFG_MX6QP) $(CFG_MX6Q) $(CFG_MX6D) $(CFG_MX6DL) $(CFG_MX6S) $(CFG_MX6ULL)), y)
+# Reference manual says: OTP Bank 4 Word 6 (HW Capabilities) (OCOTP_GP1)
+CFG_IMX_OCOTP_MANUFACTURING_BANK  ?= 4
+CFG_IMX_OCOTP_MANUFACTURING_WORD  ?= 6
+CFG_IMX_OCOTP_MANUFACTURING_BIT   ?= 0
+endif
+
+ifeq ($(filter y, $(CFG_MX93)), y)
+CFG_IMX_OCOTP_MANUFACTURING_BANK  ?= 47
+CFG_IMX_OCOTP_MANUFACTURING_WORD  ?= 0
+CFG_IMX_OCOTP_MANUFACTURING_BIT   ?= 0
+endif
+
+ifeq ($(filter y, $(CFG_MX8M)), y)
+CFG_IMX_OCOTP_MANUFACTURING_BANK  ?= 14
+CFG_IMX_OCOTP_MANUFACTURING_WORD  ?= 0
+CFG_IMX_OCOTP_MANUFACTURING_BIT   ?= 0
+endif
+
 ifneq (,$(filter $(PLATFORM_FLAVOR),mx7dsabresd))
 CFG_DDR_SIZE ?= 0x40000000
 CFG_NS_ENTRY_ADDR ?= 0x80800000
diff --git a/core/arch/arm/plat-imx/main.c b/core/arch/arm/plat-imx/main.c
index 500505074d78..e56260cf2ec7 100644
--- a/core/arch/arm/plat-imx/main.c
+++ b/core/arch/arm/plat-imx/main.c
@@ -32,12 +32,14 @@
 #include <console.h>
 #include <drivers/gic.h>
 #include <drivers/imx_uart.h>
+#include <drivers/imx_ocotp.h>
 #include <imx.h>
 #include <kernel/boot.h>
 #include <mm/core_memprot.h>
 #include <mm/core_mmu.h>
 #include <platform_config.h>
 #include <stdint.h>
+#include <pta_manufacturing.h>
 
 static struct imx_uart_data console_data __nex_bss;
 
@@ -128,3 +130,38 @@ void boot_secondary_init_intc(void)
 	gic_init_per_cpu();
 }
 #endif
+
+#if defined(CFG_MANUFACTURING_PTA) &&                 \
+	defined(CFG_IMX_OCOTP_MANUFACTURING_BANK) &&  \
+	defined(CFG_IMX_OCOTP_MANUFACTURING_WORD) &&  \
+	defined(CFG_IMX_OCOTP_MANUFACTURING_BIT)
+
+TEE_Result pta_manufacturing_query_state(enum pta_manufacturing_state *state)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t val = 0;
+
+	res = imx_ocotp_read(CFG_IMX_OCOTP_MANUFACTURING_BANK,
+			     CFG_IMX_OCOTP_MANUFACTURING_WORD, &val);
+	if (res) {
+		EMSG("Failed to read manufacturing fuse.");
+		*state = PTA_MANUFACTURING_STATE_UNKNOWN;
+		return res;
+	}
+
+	val = (val >> CFG_IMX_OCOTP_MANUFACTURING_BIT) & 1U;
+
+	*state = (enum pta_manufacturing_state)val;
+	return TEE_SUCCESS;
+}
+
+TEE_Result pta_manufacturing_set_state(enum pta_manufacturing_state state)
+{
+	uint32_t val = state;
+
+	val = (val & 1U) << CFG_IMX_OCOTP_MANUFACTURING_BIT;
+
+	return imx_ocotp_write(CFG_IMX_OCOTP_MANUFACTURING_BANK,
+			       CFG_IMX_OCOTP_MANUFACTURING_WORD, val);
+}
+#endif
